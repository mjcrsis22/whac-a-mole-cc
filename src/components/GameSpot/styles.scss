.game-spot {
  position: relative;
  width: 2.5rem;
  height: 3rem;

  .spot-mark {
    position: absolute;
    left: 0;
    top: 1.5rem;
    width: 2.5rem;
    height: 1.5rem;
    border-radius: 50%;
    background-color: #583806;
    transition-property: transform;
    transition-duration: 300ms;
    transform: scale(0);

    &.ready {
      transform: scale(1);
    }
  }

  .spot-mole-wrapper {
    position: absolute;
    left: 0;
    top: 0;
    width: 100%;
    height: 2.5rem;
    overflow: hidden;

    .spot-mole {
      position: absolute;
      left: 0;
      top: 2.5rem;
      width: 100%;
      transition-property: top;
      transition-duration: 500ms;

      &.show {
        top: 0;
      }
    }
  }
}
