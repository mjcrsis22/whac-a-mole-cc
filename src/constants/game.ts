export enum GAME_LEVEL {
  LOW = 'LOW',
  MEDIUM = 'MEDIUM',
  HARD = 'HARD',
}

export interface GameSetting {
  id: GAME_LEVEL;
  spotRate: number;
  points: number;
}

export const GAME_SETTINGS_BY_LEVEL: {
  [key in keyof typeof GAME_LEVEL]: GameSetting;
} = {
  [GAME_LEVEL.LOW]: {
    id: GAME_LEVEL.LOW,
    spotRate: 1000,
    points: 10,
  },
  [GAME_LEVEL.MEDIUM]: {
    id: GAME_LEVEL.MEDIUM,
    spotRate: 750,
    points: 20,
  },
  [GAME_LEVEL.HARD]: {
    id: GAME_LEVEL.HARD,
    spotRate: 500,
    points: 30,
  },
};
