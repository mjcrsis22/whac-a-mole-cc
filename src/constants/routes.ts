export enum APP_ROUTES {
  HOME = '/home',
  GAME = '/game',
}
