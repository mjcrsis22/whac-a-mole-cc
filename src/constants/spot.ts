export enum SPOT_STATES {
  inactive,
  active,
}

export const SPOT_POSITIONS = [
  { left: '5%', top: '0%' },
  { left: '5%', top: '72%' },
  { left: '30%', top: '15%' },
  { left: '30%', top: '60%' },
  { left: '45%', top: '40%' },
  { left: '60%', top: '15%' },
  { left: '60%', top: '60%' },
  { left: '85%', top: '0%' },
  { left: '85%', top: '72%' },
];
