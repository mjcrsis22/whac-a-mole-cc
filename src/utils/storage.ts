export enum STORE_KEYS {
  USERNAME = 'wam_username',
}

export const storeData = (key: string, value: string): void => {
  localStorage.setItem(key, value);
};

export const getStoredData = (key: string): string | null => {
  return localStorage.getItem(key);
};

export const removeStoredData = (key: string): void => {
  localStorage.removeItem(key);
};
