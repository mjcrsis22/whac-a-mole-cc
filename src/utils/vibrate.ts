export const vibrate = (pattern: VibratePattern): boolean => {
  return window.navigator?.vibrate?.(pattern);
};
