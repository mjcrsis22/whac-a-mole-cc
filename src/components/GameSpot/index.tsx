import { useContext, useEffect, useState } from 'react';
import { GameContext } from '../../contexts/GameContext';
import { ActionTypes } from '../../contexts/GameContext/reducer';
import { SPOT_STATES } from '../../constants/spot';
import { vibrate } from '../../utils/vibrate';
import moleLogo from '../../assets/mole.png';
import './styles.scss';

interface GameSpotProps {
  state: SPOT_STATES;
}

function GameSpot({ state }: GameSpotProps) {
  const {
    state: { points, levelSetting },
    dispatch,
  } = useContext(GameContext);
  const [spotReady, setSpotReady] = useState(false);
  const [showMole, setShowMole] = useState(false);

  const moleClick = () => {
    vibrate(200);
    dispatch({
      type: ActionTypes.SET_POINTS,
      payload: points + levelSetting.points,
    });
  };

  useEffect(() => {
    window.setTimeout(() => setSpotReady(true), 100);
    return () => {
      setSpotReady(false);
    };
  }, []);

  useEffect(() => {
    setShowMole(state == SPOT_STATES.active);
  }, [state]);

  return (
    <div className="game-spot">
      <div className={`spot-mark ${spotReady ? 'ready' : ''}`}></div>
      <div className="spot-mole-wrapper">
        <img
          className={`spot-mole ${showMole ? 'show' : ''}`}
          src={moleLogo}
          alt="Mole"
          role="button"
          onClick={moleClick}
        />
      </div>
    </div>
  );
}

export default GameSpot;
