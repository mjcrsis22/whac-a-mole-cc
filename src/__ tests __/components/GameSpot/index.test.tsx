import { useContext } from 'react';
import { render, act, fireEvent } from '@testing-library/react';
import GameContextProvider from '../../../contexts/GameContext';
import {
  ActionTypes,
  INITIAL_STATE,
} from '../../../contexts/GameContext/reducer';
import { vibrate } from '../../../utils/vibrate';
import { SPOT_STATES } from '../../../constants/spot';
import GameSpot from '../../../components/GameSpot';

jest.mock('react', () => ({
  ...jest.requireActual('react'),
  useContext: jest.fn(),
}));

jest.mock('../../../utils/vibrate', () => ({
  ...jest.requireActual('../../../utils/vibrate'),
  vibrate: jest.fn(),
}));

const mockedContext = {
  state: INITIAL_STATE,
  dispatch: jest.fn(),
};

const mockedUseContext = useContext as jest.MockedFn<typeof useContext>;

describe('GameSpot component', () => {
  beforeAll(() => {
    jest.useFakeTimers();
  });

  beforeEach(() => {
    mockedUseContext.mockReturnValue(mockedContext);
  });

  afterEach(() => {
    jest.clearAllMocks();
  });

  test.each([SPOT_STATES.inactive, SPOT_STATES.active])(
    'Renders correctly - according to the state prop',
    async stateProp => {
      const { container } = render(
        <GameContextProvider>
          <GameSpot state={stateProp} />
        </GameContextProvider>,
      );
      await act(() => {
        jest.runAllTimers();
      });
      expect(container).toMatchSnapshot();
    },
  );

  test('Dispatches SET_POINTS action on image click', async () => {
    const currentPoints = mockedContext.state.points;
    const levelPoints = mockedContext.state.levelSetting.points;
    const finalPoints = currentPoints + levelPoints;

    const result = render(
      <GameContextProvider>
        <GameSpot state={SPOT_STATES.active} />
      </GameContextProvider>,
    );

    await act(() => {
      jest.runAllTimers();
    });

    const button = result.getByRole('button');
    fireEvent(button, new MouseEvent('click', { bubbles: true }));

    expect(result.container).toMatchSnapshot();
    expect(vibrate).toHaveBeenCalledWith(200);
    expect(mockedContext.dispatch).toHaveBeenCalledWith({
      type: ActionTypes.SET_POINTS,
      payload: finalPoints,
    });
  });
});
