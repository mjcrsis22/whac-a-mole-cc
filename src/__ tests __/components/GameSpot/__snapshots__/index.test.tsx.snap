// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`GameSpot component Dispatches SET_POINTS action on image click 1`] = `
<div>
  <div
    class="game-spot"
  >
    <div
      class="spot-mark ready"
    />
    <div
      class="spot-mole-wrapper"
    >
      <img
        alt="Mole"
        class="spot-mole show"
        role="button"
        src="test-file-stub"
      />
    </div>
  </div>
</div>
`;

exports[`GameSpot component Renders correctly - according to the state prop 1`] = `
<div>
  <div
    class="game-spot"
  >
    <div
      class="spot-mark ready"
    />
    <div
      class="spot-mole-wrapper"
    >
      <img
        alt="Mole"
        class="spot-mole "
        role="button"
        src="test-file-stub"
      />
    </div>
  </div>
</div>
`;

exports[`GameSpot component Renders correctly - according to the state prop 2`] = `
<div>
  <div
    class="game-spot"
  >
    <div
      class="spot-mark ready"
    />
    <div
      class="spot-mole-wrapper"
    >
      <img
        alt="Mole"
        class="spot-mole show"
        role="button"
        src="test-file-stub"
      />
    </div>
  </div>
</div>
`;
